distributed/tests/test_scheduler.py

```diff
@@ -4475,3 +4475,84 @@ async def test_scatter_creates_ts(c, s, a, b):
         await a.close()
         assert await x2 == 2
     assert s.tasks["x"].run_spec is not None
+
+
+try:
+    from dask.delayed import single_key
+except ImportError:
+    single_key = first
+from dask.typing import DaskGraph
+
+from distributed.scheduler import _ALLOW_MATERIALIZATION
+
+
+def mysum(x, y):
+    return x + y
+
+
+class MyCustomScalar(DaskGraph):
+    dsk = {
+        "x": (inc, 1),
+        "y": (inc, "x"),
+        "z": (mysum, "x", "y"),
+    }
+    out_keys = ["z"]
+
+    # This API call is a bit ambiguous. For current dask collections this is
+    # used as "Give me the HLG if there is any, else give me a dict"
+    # This only makes sense on a "Collection" that is not a graph.
+
+    # This also poses other problems, namely it is **called** in
+    # dask.is_dask_collection to infer whether the object is an collection or
+    # not.
+    # However, dask-expr is implementing this API s.t. it returns the
+    # materialized graph.
+    # To disambiguate, I introduced `__dask_graph_nomat__` to flag places in the
+    # code that *must not* trigger accidental materialization
+
+    def __dask_graph__(self):
+        if _ALLOW_MATERIALIZATION.get():
+            return self.dsk
+        else:
+            raise RuntimeError("Not allowed to materialize!")
+
+    def __dask_graph_nomat__(self):
+        return self
+
+    def __dask_annotations__(self):
+        return {"x": {"foo": "bar"}, "z": {"retries": 1}}
+
+    def __dask_keys__(self):
+        return self.out_keys
+
+    # Everything below this is only necessary for
+    # postcompute/postpersist/finalize foo
+
+    def __dask_postcompute__(self):
+        # NOTE: first is dask.single_key which is a sentinel for client.compute
+        return single_key, ()
```

Comment:
this should rather be replaced by a `postcompute` method on the expression that performs the necessary graph manipulation itself.